Sources/ShellOut.swift

```diff
@@ -416,6 +444,91 @@ private extension Process {
             return outputData.shellOutput()
         }
     }
+    
+    func launchBash(with outputHandle: FileHandle? = nil, errorHandle: FileHandle? = nil, withCompletion completion: @escaping Completion) {
+        
+        var outputData = Data()
+        var errorData = Data()
+        
+        let outputPipe = Pipe()
+        standardOutput = outputPipe
+        
+        let errorPipe = Pipe()
+        standardError = errorPipe
+        
+        // Because FileHandle's readabilityHandler might be called from a
+        // different queue from the calling queue, avoid a data race by
+        // protecting reads and writes to outputData and errorData on
+        // a single dispatch queue.
+        let outputQueue = DispatchQueue(label: "bash-output-queue")
+        
+        #if !os(Linux)
+        outputPipe.fileHandleForReading.readabilityHandler = { handler in
+            outputQueue.async {
+                let data = handler.availableData
+                outputData.append(data)
+                DispatchQueue.main.async {
+                    outputHandle?.write(data)
+                }
+            }
+        }
+        
+        errorPipe.fileHandleForReading.readabilityHandler = { handler in
+            outputQueue.async {
+                let data = handler.availableData
+                errorData.append(data)
+                DispatchQueue.main.async {
+                    errorHandle?.write(data)
+                }
+            }
+        }
+        #endif
+        
+        launch()
+        
+        #if os(Linux)
+        outputQueue.sync {
+            outputData = outputPipe.fileHandleForReading.readDataToEndOfFile()
+            errorData = errorPipe.fileHandleForReading.readDataToEndOfFile()
+        }
+        #endif
+        
+        waitUntilExit()
+        
+        DispatchQueue.main.async {
+            outputHandle?.closeFile()
+            errorHandle?.closeFile()
+        }
+        
+        #if !os(Linux)
+        outputPipe.fileHandleForReading.readabilityHandler = nil
+        errorPipe.fileHandleForReading.readabilityHandler = nil
+        #endif
+        
+        do {
+            // Block until all writes have occurred to outputData and errorData,
+            // and then read the data back out.
+            return try outputQueue.sync {
```

Comment:
Discussion: Making sure it's dispatched to the end of the main queue, should be fine.